src/NHibernate/Engine/AssociationKey.cs

```diff
@@ -40,7 +44,22 @@ public override bool Equals(object that)
 
 		public override int GetHashCode()
 		{
-			return hashCode;
+			// If the object is put in a set or dictionary during deserialization, the hashcode will not yet be
+			// computed. Compute the hashcode on the fly. So long as this happens only during deserialization, there
+			// will be no thread safety issues. For the hashcode to be always defined after deserialization, the
+			// deserialization callback is used.
+			return _hashCode ?? GenerateHashCode();
+		}
+
+		/// <inheritdoc />
+		public void OnDeserialization(object sender)
+		{
+			_hashCode = GenerateHashCode();
+		}
+
+		private int GenerateHashCode()
+		{
+			return ownerKey.GetHashCode() ^ propertyName.GetHashCode() ^ ownerKey.EntityName.GetHashCode();
```

Comment:
Maybe for apparently simple cases like this one we would have better to always compute it on the fly instead of caching it. (`ownerKey.GetHashCode` computes a bit more than the two other ones which are just `string.GetHashCode()`.)

Reply:
Please always compute this.